### PR TITLE
Add CommandPropertyToAttributeRector

### LINKED ISSUE #159: Feature idea: Add the Symfony\Component\Console\Attribute\AsCommand attribute
As of Symfony 6.1, I've got this notification in my logs:

> Since symfony/console 6.1: Relying on the static property "$defaultDescription" for setting a command description is deprecated. Add the "Symfony\Component\Console\Attribute\AsCommand" attribute to the "App\Command\NotificationCommand" class instead.

I'd the great if Rector can refactor commands with `static $defaultName` and `static $defaultDescription`.

### PR DESCRIPTION
**Pull request by stephanvierkant**
My first PR that adds a Rector, so happy to get some feedback.

Fixes #159

### CODE REVIEW

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+     * @var \Rector\Php80\NodeAnalyzer\PhpAttributeAnalyzer
+     */
+    private $phpAttributeAnalyzer;
+    /**
+     * @readonly
+     * @var \Rector\PhpAttribute\Printer\PhpAttributeGroupFactory
+     */
+    private $phpAttributeGroupFactory;
+
+    public function __construct(\Rector\PhpAttribute\Printer\PhpAttributeGroupFactory $phpAttributeGroupFactory, \Rector\Php80\NodeAnalyzer\PhpAttributeAnalyzer $phpAttributeAnalyzer)
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:29]: property promotion and readonly can be used here. You can run:

```
vendor/bin/rector process src/Rector/Class_/CommandPropertyToAttributeRector.php
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * @see \Rector\Symfony\Tests\Rector\Class_\CommandPropertyToAttributeRector\CommandPropertyToAttributeRectorTest
+ */
+final class CommandPropertyToAttributeRector extends AbstractRector implements MinPhpVersionInterface
+{
+    private const ATTRIBUTE = 'Symfony\Component\Console\Attribute\AsCommand';
+
+    public function __construct(
+        /**
+         * @readonly
+         */
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:31]: docblock readonly no longer needed

```suggestion
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node) : ?Node
+    {
+        if (!$this->isObjectType($node, new ObjectType('Symfony\\Component\\Console\\Command\\Command'))) {
+            return null;
+        }
+
+        if (!class_exists(self::ATTRIBUTE)) {
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:78]: it should use reflectionProvider instead:

```suggestion
        if (! $this->reflectionProvider->hasClass(self::ATTRIBUTE)) {
```

You can inject dependency to `__construct()`:

```php
private readonly \PHPStan\Reflection\ReflectionProvider $reflectionProvider
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            return null;
+        }
+
+        if ($this->phpAttributeAnalyzer->hasPhpAttribute($node, self::ATTRIBUTE)) {
+            return null;
+        }
+
+        $defaultName = null;
+        $defaultDescription = null;
+
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:88]: when `count($node->props) > 1`, it should be skipped:

```php
if (count($node->props) > 1) {
    return null;
}
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        $property = $node->getProperty('defaultName');
+        if ($property !== null) {
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:90]:
```suggestion
        if ($property instanceof \PhpParser\Node\Stmt\Property) {
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            $propertyProperty = $property->props[0];
+            if ($propertyProperty->default instanceof String_) {
+                $defaultName = $propertyProperty->default->value;
+            }
+
+            $this->removeNode($property);
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:96]: `removeNode()` should only be called when default value is string, so check need to be in if above I guess.

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        $property = $node->getProperty('defaultDescription');
+        if ($property !== null) {
+            $propertyProperty = $property->props[0];
+            if($propertyProperty->default instanceof String_) {
+                $defaultDescription = $propertyProperty->default->value;
+            }
+            $this->removeNode($property);
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:105]: `removeNode()` should only be called when default value is string, so check need to be in if above I guess.

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:107]: `defaultName` and `defaultDescription` need to have initialization variable, so below, we can check if `defaultName` and `defaultDescription` is not null / empty based on initialiazation.

File: src/Rector/Class_/CommandPropertyToAttributeRector.php (hunk 4 above)
Comment (samsonasik) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:85]:
```suggestion
    if (count($node->props) !== 1) {
        return null;
    }
```

Comment (stephanvierkant) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:93]: I don't get this. Why is this needed?

Comment (samsonasik) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:93]: property can be defined as `private $a, $b, $c;`

**Comment by TomasVotruba** [comment]
I've triggered CI :+1:

**Comment by TomasVotruba** [comment]
I've triggered CI :+1:

File: src/Rector/Class_/CommandPropertyToAttributeRector.php (hunk 4 above)
Comment (stephanvierkant) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:94]: And why would this be a problem?

Comment (samsonasik) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:94]: removing part of property should not remove whole props if there is other props exists, that may be used, as this rule remove the property itself.

Comment (stephanvierkant) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:94]: Thanks for clarification. I'll fix it!

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        $property = $node->getProperty('defaultDescription');
+        if ($property instanceof Property) {
+            $propertyProperty = $property->props[0];
+            if ($propertyProperty->default instanceof String_) {
+                $defaultDescription = $propertyProperty->default->value;
+
+                $this->removeNode($property);
+            }
+        }
+
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:118]: 1 more check:

```suggestion
       if ($defaultDescription === null) {
            return null;
        }
```

Comment (stephanvierkant) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:115]: Description isn't required, is it?

Comment (samsonasik) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:115]: you're correct 👍

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you :+1:  :clap:

**Pull request merged**